backend/src/main/java/botobo/core/application/WorkbookService.java

```diff
@@ -86,8 +93,43 @@ public WorkbookCardResponse findWorkbookCardsById(Long id) {
 
     @Transactional
     public void deleteWorkbook(Long id, AppUser appUser) {
-        Workbook workbook = workbookRepository.findById(id)
-                .orElseThrow(WorkbookNotFoundException::new);
+        Workbook workbook = findWorkbook(id);
         workbook.deleteIfUserIsAuthor(appUser.getId());
     }
+
+    @Transactional
+    public void scrapSelectedCardsToWorkbook(Long workbookId, ScrapCardRequest scrapCardRequest, AppUser appUser) {
+        User user = findUser(appUser);
+        Workbook workbook = findWorkbook(workbookId);
+        if (!workbook.isAuthorOf(user)) {
+            throw new NotAuthorException();
+        }
+        Cards scrappedCards = new Cards(scrapCards(scrapCardRequest.getCardIds()));
+        addScrappedCardsToWorkbook(workbook, scrappedCards);
+    }
+
+    private User findUser(AppUser appUser) {
+        return userRepository.findById(appUser.getId())
+                .orElseThrow(UserNotFoundException::new);
+    }
+
+    private Workbook findWorkbook(Long workbookId) {
+        return workbookRepository.findById(workbookId)
+                .orElseThrow(WorkbookNotFoundException::new);
+    }
+
+    private List<Card> scrapCards(List<Long> cardIds) {
+        List<Card> scrappedCards = new ArrayList<>();
+        for (Long id : cardIds) {
+            final Card sourceCard = cardRepository.findById(id)
+                    .orElseThrow(CardNotFoundException::new);
+
+            scrappedCards.add(Card.createCopyOf(sourceCard));
+        }
+        return scrappedCards;
+    }
```

Comment:
```suggestion
    private List<Card> scrapCards(List<Long> cardIds) {
        List<Card> foundCards = cardRepository.findByIdIn(cardIds);
        if (!Objects.equals(cardIds.size(), foundCards.size())) {
            throw new CardNotFoundException();
        }
        return foundCards.stream()
                .map(Card::createCopyOf)
                .collect(Collectors.toList());
    }
```
쉬는 날이라 심심해서 한 번 리팩토링 해봤어요. 😁
여기서 바뀐점은 크게 두 가지입니다:
1. 기존에는 카드 한 장씩 select를 해서 가지고 왔다 -> 한 번의 쿼리로 일단 카드가 있는지 없는지 상관하지 않고 다 갖고 온다.
2. 해당 카드 List 의 사이즈를 카드 id List 사이즈와 비교해서 그 크기가 같지 않다면 CardNotFound 예외를 던진다.

이게 가능한 이유가 CardNotFoundException이 터졌을 때 어떤 카드가 없는지 메시지에 포함시키지 않아서 입니다.
스트림 부분은 딱히 바꿀 필요는 없지만 쿼리가 한 번만 나가게 하는건 이 방식이 괜찮은 것 같아요.
조앤은 어떻게 생각하시는지 궁금하네요.

만약 이렇게 바꾸면 AcceptanceTest 는 바꿀 필요없지만 ServiceTest에는 findById 가 호출되었는지 확인하는 부분을 findByIdIn 으로 바꿔줘야합니다!

Reply:
오..!! 피케이 아이디어 좋은거 같은데요??
아무래도 쿼리가 여러번 나가는건 부담이 많이가니까요!! 
하나씩 find하면서 예외처리 안하고 한번에 불러서 예외처리해줘도 괜찮다고 생각이 드네요!!

Reply:
그러게요 천재인 듯;;